internal/provider/resource_credential.go

```diff
@@ -6,209 +6,307 @@ import (
+	"github.com/hashicorp/terraform-plugin-framework/resource/schema/booldefault"
+	"github.com/hashicorp/terraform-plugin-framework/resource/schema/objectplanmodifier"
+	"github.com/hashicorp/terraform-plugin-framework/resource/schema/planmodifier"
+	"github.com/hashicorp/terraform-plugin-framework/resource/schema/stringplanmodifier"
+	"github.com/hashicorp/terraform-plugin-framework/types"
+	"github.com/hashicorp/terraform-plugin-framework/types/basetypes"
+
+	"github.com/hashicorp/terraform-plugin-log/tflog"
 	"github.com/juju/terraform-provider-juju/internal/juju"
 )
 
-func resourceCredential() *schema.Resource {
-	return &schema.Resource{
-		Description: "A resource that represent a credential for a cloud.",
+// Ensure provider defined types fully satisfy framework interfaces.
+var _ resource.Resource = &credentialResource{}
+var _ resource.ResourceWithConfigure = &credentialResource{}
+var _ resource.ResourceWithImportState = &credentialResource{}
 
-		CreateContext: resourceCredentialCreate,
-		ReadContext:   resourceCredentialRead,
-		UpdateContext: resourceCredentialUpdate,
-		DeleteContext: resourceCredentialDelete,
-		Importer: &schema.ResourceImporter{
-			StateContext: resourceCredentialImporter,
-		},
+func NewCredentialResource() resource.Resource {
+	return &credentialResource{}
+}
 
-		Schema: map[string]*schema.Schema{
-			"cloud": {
+type credentialResource struct {
+	client *juju.Client
+}
+
+type credentialResourceModel struct {
+	Cloud                types.Object `tfsdk:"cloud"`
+	Attributes           types.Map    `tfsdk:"attributes"`
+	AuthType             types.String `tfsdk:"auth_type"`
+	ClientCredential     types.Bool   `tfsdk:"client_credential"`
+	ControllerCredential types.Bool   `tfsdk:"controller_credential"`
+	Name                 types.String `tfsdk:"name"`
+
+	// ID required by the testing framework
+	ID types.String `tfsdk:"id"`
+}
+
+func (c *credentialResource) Metadata(_ context.Context, req resource.MetadataRequest, resp *resource.MetadataResponse) {
+	resp.TypeName = req.ProviderTypeName + "_credential"
+}
+
+func (c *credentialResource) Schema(_ context.Context, _ resource.SchemaRequest, resp *resource.SchemaResponse) {
+	resp.Schema = schema.Schema{
+		Description: "A resource that represent a credential for a cloud.",
+		Blocks: map[string]schema.Block{
+			"cloud": schema.SingleNestedBlock{
 				Description: "JuJu Cloud where the credentials will be used to access",
-				Type:        schema.TypeList,
-				Optional:    true,
-				Computed:    true,
-				ForceNew:    true,
-				MaxItems:    1,
-				Elem: &schema.Resource{
-					Schema: map[string]*schema.Schema{
-						"name": {
-							Description: "The name of the cloud",
-							Type:        schema.TypeString,
-							Required:    true,
-						},
+				Attributes: map[string]schema.Attribute{
+					"name": schema.StringAttribute{
+						Description: "The name of the cloud",
+						Required:    true,
```

Comment:
those are a bit different in Blocks (as opposed to Attributes). I believe all the Blocks are optional. I'm not sure about the computed bit, I think it's not a thing for `ListNestedBlock`s, but it is for `ListNestedAttributes` (which we can't use because `cloud = {}` vs `cloud {}`)

https://developer.hashicorp.com/terraform/plugin/framework/migrating/attributes-blocks/blocks

https://developer.hashicorp.com/terraform/plugin/framework/migrating/attributes-blocks/blocks-computed